lib/api/CBoostedTreeInferenceModelBuilder.cc

```diff
@@ -0,0 +1,193 @@
+    if (inputColumnIndex < m_Definition.input().fieldNames().size()) {
+        std::string fieldName{m_Definition.input().fieldNames()[inputColumnIndex]};
+        std::string featureName{fieldName + "_targetmean"};
+        auto stringMap = this->encodingMap(inputColumnIndex, map);
+        m_Definition.preprocessors().push_back(std::make_unique<CTargetMeanEncoding>(
+            fieldName, fallback, featureName, std::move(stringMap)));
+        m_FeatureNames.push_back(featureName);
+    }
+}
+
+void CBoostedTreeInferenceModelBuilder::addFrequencyEncoding(std::size_t inputColumnIndex,
+                                                             const TDoubleVec& map) {
+    if (inputColumnIndex < m_Definition.input().fieldNames().size()) {
+        std::string fieldName{m_Definition.input().fieldNames()[inputColumnIndex]};
+        std::string featureName{fieldName + "_frequency"};
+        auto stringMap = this->encodingMap(inputColumnIndex, map);
+        m_Definition.preprocessors().push_back(std::make_unique<CFrequencyEncoding>(
+            fieldName, featureName, std::move(stringMap)));
+        m_FeatureNames.push_back(featureName);
+    }
+}
+
+CInferenceModelDefinition&& CBoostedTreeInferenceModelBuilder::build() {
+
+    // Finalize OneHotEncoding Mappings
+    for (auto& oneHotEncodingMapping : m_OneHotEncodingMaps) {
+        m_Definition.preprocessors().emplace_back(
+            std::move(oneHotEncodingMapping.second));
+    }
+
+    // Add aggregated output after the number of trees is known
+    auto ensemble{static_cast<CEnsemble*>(m_Definition.trainedModel().get())};
+    this->setAggregateOutput(ensemble);
+
+    this->setTargetType();
+    ensemble->featureNames(m_FeatureNames);
+
+    return std::move(m_Definition);
+}
+
+void CBoostedTreeInferenceModelBuilder::addNode(
+    std::size_t splitFeature,
+    double splitValue,
+    bool assignMissingToLeft,
+    double nodeValue,
+    double gain,
+    ml::maths::CBoostedTreeNode::TOptionalNodeIndex leftChild,
+    ml::maths::CBoostedTreeNode::TOptionalNodeIndex rightChild) {
+    auto ensemble{static_cast<CEnsemble*>(m_Definition.trainedModel().get())};
+    // use dynamic cast to prevent using wrong type of trained models
+    auto tree = dynamic_cast<CTree*>(ensemble->trainedModels().back().get());
+    if (tree == nullptr) {
+        HANDLE_FATAL(<< "Internal error. Tree points to a nullptr.")
+    }
+    tree->treeStructure().emplace_back(tree->size(), splitValue, assignMissingToLeft, nodeValue,
+                                       splitFeature, leftChild, rightChild, gain);
+}
+
+CBoostedTreeInferenceModelBuilder::CBoostedTreeInferenceModelBuilder(
+    TStringVec fieldNames,
+    std::size_t dependentVariableColumnIndex,
+    const TStringSizeUMapVec& categoryNameMap) {
+    // filter filed names containing only "."
+    fieldNames.erase(std::remove(fieldNames.begin(), fieldNames.end(), "."),
+                     fieldNames.end());
```

Comment:
The field names I store on the data frame do not include the special control fields. So I think this is redundant.

Reply:
It's not to filter control fields, but the additional columns in the data frame after the target column:
```    TStrVec fieldNames{"numeric_col", "categorical_col", "target_col", ".", "."};```

Reply:
Sorry these are what I was referring to: one is the field we use to pass control messages from the Java. I explicitly don't copy them to the data frame see [this](https://github.com/elastic/ml-cpp/blob/master/lib/api/CDataFrameAnalyzer.cc#L250) line.

Reply:
So my question would be are you sure this is still the case, i.e. I changed this recently. If you're somehow getting them still we should probably investigate.

Reply:
Talked about this offline and it seems like they might still be there. This needs investigation, but shouldn't hold this PR up.